### ISSUE TITLE
catalog, zen_image: handling missing, defined image

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4112]
I note that there is no substitution if an image is defined, but missing:
https://github.com/zencart/zencart/blob/6a4ac062128866d02819fac565cc517a15be8746/includes/functions/html_output.php#L185-L197

The image link produced will still point to the missing image.

This snippet seems like it may benefit from a refactor in this respect, as checks are made for DIR_WS_IMAGES, but when no image is defined, $src is empty.

**Comment by drbyte** [comment]
We could add a `file_exists()` check for `DIR_WS_IMAGES . PRODUCTS_IMAGE_NO_IMAGE`.

Two caveats: 
- that would mean another hit to the filesystem (speed)
- it further embeds image handling to be supported only for the local server. (Yes, DIR_WS_IMAGES already means it's gotta be local, not remote, but adding a file_exists check further cements local-only capability.)

Under what circumstance are you encountering that PRODUCTS_IMAGE_NO_IMAGE is defined but missing? Who's deleting the image?

**Comment by torvista** [comment]
I'm commenting on the case when a manufacturer/category/product has a real image defined/specified, but it no longer exists/has been moved
I may get this when spring-cleaning/having to do some reworking of the organisation/renaming/splitting of image folders, as time goes on, and not being as thorough as necessary when reworking the links in the db.
I feel I have my images under-control, but I imagine most users do not initially and inevitably face some re-organisation at some point, when this issue may arise.
Also when working on an upgrade, I do not duplicate the majority of images onto the new site in the interim, so most are missing.
I just thought it would be correct practice to prevent broken links, for whatever reason, since it looks bad.
In fact I create a debug for missing images such as this.

**Comment by torvista** [comment]
And regarding the filesystem hit, there is already a test for the image existing but it fails all images, always:
```
// if not in current template switch to template_default 
     if (!file_exists($src)) { 
       $src = str_replace(DIR_WS_TEMPLATES . $template_dir, DIR_WS_TEMPLATES . 'template_default', $src); 
     } 
```
as $src is not a complete path, for example "/images/folder/product1_image1.jpg".

I don't really get that blind substitution...

**Comment by drbyte** [comment]
> I'm commenting on the case when a manufacturer/category/product has a real image defined/specified, but it no longer exists/has been moved

I misread your post because you talked about "defined" and I immediately thought of "constants" and the first lines of the code snippet you quoted referred to the PRODUCTS_IMAGE_NO_IMAGE constant that gets defined via admin config.

You are correct: if the database contains a string that should identify an image filename (instead of being blank) then no further checking is done to substitute something else if the file does not exist.

**Comment by drbyte** [comment]
> I don't really get that blind substitution...

?

**Comment by torvista** [comment]
` $src = str_replace(DIR_WS_TEMPLATES . $template_dir, DIR_WS_TEMPLATES . 'template_default', $src); `
If the file does not exist, this is attempted, "blind" because the new src may not exist either.

I did this for now,
```
$src = str_replace(DIR_WS_TEMPLATES . $template_dir, DIR_WS_TEMPLATES . 'template_default', $src, $replaced);
        if ($replaced === 0) {
            $src = DIR_WS_IMAGES . PRODUCTS_IMAGE_NO_IMAGE;        }
```

**Comment by torvista** [comment]
I had to document this process....when I explain something, I understand better myself, so...regarding the image defined for a product that does not exist (moved, renamed or just not existing in a development site to save space).

Main Product Image
Product image is defined in db as ht_gpdt/gpdt.jpg

Processing begins in /templates/tpl_modules_main_product_image.php, 
then modules/main_product_image.php (required module).

1) These variables are created:
$products_image_extension = substr($products_image, strrpos($products_image, '.'));
$products_image_base = str_replace($products_image_extension, '', $products_image);
$products_image_medium = $products_image_base . IMAGE_SUFFIX_MEDIUM . $products_image_extension;
$products_image_large = $products_image_base . IMAGE_SUFFIX_LARGE . $products_image_extension;

eg.
Values BEFORE being handled:
$products_image=ht_gpdt/gpdt.jpg
$products_image_extension=.jpg
$products_image_base=ht_gpdt/gpdt
$products_image_medium=ht_gpdt/gpdt_MED.jpg
$products_image_large=ht_gpdt/gpdt_LRG.jpg

2) Next step
a) Default/if NOT handled

Checks are made for existence of _MED and _LRG versions, and if they exist, load those names into the respective variables.

In the case of using IH, those _MED and _LRG do not exist, so if IH is disabled, those variables are set to the base image:
$main_image_handled=
$products_image=ht_gpdt/gpdt.jpg
$products_image_extension=.jpg
$products_image_base=ht_gpdt/gpdt
$products_image_medium=images/ht_gpdt/gpdt.jpg
$products_image_large=images/ht_gpdt/gpdt.jpg

b) If handled: IH observer uses NOTIFY_MODULES_MAIN_PRODUCT_IMAGE_FILENAME

The observer removes any extension from the $products_image to load the $products_image_base with the stub name.
Not sure why, as this is already done in core code earlier??

It also adds the medium and large image names to the respective variables as though they DO exist (although they do not exist).

Values AFTER being handled:
$main_image_handled=1
$products_image=ht_gpdt/gpdt.jpg
$products_image_extension=.jpg
$products_image_base=ht_gpdt/gpdt
$products_image_medium=images/medium/ht_gpdt/gpdt_MED.jpg
$products_image_large=images/large/ht_gpdt/gpdt_LRG.jpg

3) Continuing to the generation of the main image src and link to larger image

In function zen_image, there is a hook for handle_image which passes the image to that function if the function exists (whether the plugin is enabled or not...).

a) Default/image is not handled (because function handle_image does not exist or IH is not enabled)
images src,  uses zen_image(addslashes($products_image_medium...which will be resized by the browser as per MEDIUM_IMAGE_WIDTH, MEDIUM_IMAGE_HEIGHT

image a href link,  uses javascript popup window. This has its own processing to display a larger image.

OR

b) Image is handled
image src: functions_bmz_image_handler.php, function handle_image processes the non-existing filename _MED and replaces it with a resized version which is passed back to zen_image:
bmz_cache/ht_gpdt/gpdtjpg.image.200x200.jpg


Handling missing images is more complicated than I expected since filenames are passed to IH that do not exist like images/medium/ht_gpdt/gpdt_MED.jpg, that DO get handled correctly.

So, I've concluded that if a check on the full filename, then if fails, on the base name, is done prior to the handling in zen_image and a missing image is subsequently substituted, that ensures an existing image is always used by default/also sent to IH.

I've got this as I want a debug for such missing images.

```
            // steve check
      if (!file_exists($src)) {
          $src_temp = preg_replace('/' . IMAGE_SUFFIX_MEDIUM . '\./', '.', $src);
          $src_temp = preg_replace('/' . IMAGE_SUFFIX_LARGE . '\./', '.', $src_temp);
          $src_temp = str_replace(['/medium/', '/large/'], '/', $src_temp);
          if (function_exists('handle_image') || !file_exists($src_temp)) {
              $src = DIR_WS_IMAGES . PRODUCTS_IMAGE_NO_IMAGE;
              if (IMAGE_MISSING_LOG_ERROR === 'true' && (!empty($_GET['products_id']) && zen_products_lookup($_GET['products_id'], 'products_status') === '1')) {
                  error_log('function zen_image: missing image=' . $src_temp);
              }
          }
      }
//eof
    // hook for handle_image() function such as Image Handler etc
    if (function_exists('handle_image')) {
```

I do note that checks on image names are duplicated (popup image, popup additional, main_product_image) and maybe could do with a review in general.

**Comment by torvista** [comment]
I think it is still a valid issue. It'll go on the PR list.....

**Comment by lat9** [comment]
Closing due to merged/fixing PR #4894.
